src/clr-angular/utils/focus-trap/focus-trap.directive.ts

```diff
@@ -13,50 +13,99 @@ import {
+    const nativeElement: HTMLElement = this.el.nativeElement;
 
-    if (this.focusTrapsTracker.current === this && !nativeElement.contains(event.target)) {
+    if (this.focusTrapsTracker.current === this && event.target && !nativeElement.contains(event.target)) {
       nativeElement.focus();
     }
   }
 
-  ngAfterViewInit() {
-    if (isPlatformBrowser(this.platformId)) {
-      this._previousActiveElement = <HTMLElement>document.activeElement;
-      const nativeElement: HTMLElement = this.elementRef.nativeElement;
-      nativeElement.setAttribute('tabindex', '0');
+  private createFocusableOffScreenEl(nodeName: string, optionalClass?: string): any {
+    const offScreenSpan = this.renderer.createElement(nodeName);
+    this.renderer.setAttribute(offScreenSpan, 'tabindex', '0');
+    this.renderer.addClass(offScreenSpan, 'offscreen-focus-rebounder');
+
+    if (optionalClass) {
+      this.renderer.addClass(offScreenSpan, optionalClass);
+    }
+
+    return offScreenSpan;
+  }
+
+  private addReboundEls() {
+    // We will add these focus rebounding elements only in the following conditions:
+    // 1. It should be running inside browser platform as it accesses document.body element
+    // 2. We should add them more than once. Hence, we are counting a number of focus trappers
+    //    and only add on the first focus trapper.
+
+    if (this.focusTrapsTracker.nbFocusTrappers === 1) {
+      this.topReboundEl = this.createFocusableOffScreenEl('span', 'before-focus-trap');
+      this.bottomReboundEl = this.createFocusableOffScreenEl('span', 'after-focus-trap');
+
+      // Add reboundBeforeTrapEl to the document body as the first child
+      const bodyFirstChild = this.document.body.firstChild;
+      if (bodyFirstChild) {
+        this.renderer.insertBefore(this.document.body, this.topReboundEl, bodyFirstChild);
+      }
+
+      // Add reboundAfterTrapEl to the document body as the last child
+      this.renderer.appendChild(this.document.body, this.bottomReboundEl);
+    }
+  }
+
+  private removeReboundEls() {
+    if (this.focusTrapsTracker.nbFocusTrappers === 1 && this.topReboundEl && this.bottomReboundEl) {
+      this.renderer.removeChild(this.document.body, this.topReboundEl);
+      this.renderer.removeChild(this.document.body, this.bottomReboundEl);
     }
   }
 
   public setPreviousFocus(): void {
-    if (this._previousActiveElement && this._previousActiveElement.focus) {
-      this._previousActiveElement.focus();
+    if (this.previousActiveElement && this.previousActiveElement.focus) {
+      this.previousActiveElement.focus();
+    }
+  }
+
+  ngAfterViewInit() {
+    if (isPlatformBrowser(this.platformId)) {
+      this.previousActiveElement = <HTMLElement>this.document.activeElement;
+      this.renderer.setAttribute(this.el.nativeElement, 'tabindex', '0');
+
+      this.addReboundEls();
     }
   }
 
   ngOnDestroy() {
+    if (isPlatformBrowser(this.platformId)) {
```

Comment:
I like the idea of putting the platform check inside of the method that actually uses the body or nativeElement directly, instead of in a higher caller, can we move that around?

Reply:
Now checking inside the method.